File: src/main/java/com/linxiaoyan/wikimirs/MathmlTokenizer.java
package com.linxiaoyan.wikimirs;

import java.util.ArrayList;
import java.util.List;

public class MathmlTokenizer {

	public void toTokens(MathmlNode node, List<FormulaTerm> tokens) {
		toTokens(node, 1, tokens);
	}

	private void toTokens(MathmlNode node, int level, List<FormulaTerm> tokens) {
		if (node.isLeaf()) {
			if (node.isText() && node.getLabel().length() == 1) {
				return;
			}
			tokens.add(new FormulaTerm(node.toString(), level, false));
		} else if (isNaiveNode(node)) {
			String text = node.iterator().next().getLabel();
			if (text.length() != 1) {
				tokens.add(new FormulaTerm(text, level, false));
			}
			return;
		} else {
			String gene = getChildrenTagString(node);
			String nonGene = getChildrenString(node);
			boolean doNotGenerateNonGene = false;
			if (node.getLabel().equals("math")
					|| node.getLabel().equals("mrow")) {
				if (node.size() == 1) {
					if (isNaiveNode(node.iterator().next())) {
					} else {
						doNotGenerateNonGene = true;
					}
				}
			} else {
				gene = getTagString(node, gene);
				nonGene = getTagString(node, nonGene);
			}
			if (!gene.equals(nonGene)) {
				tokens.add(new FormulaTerm(gene, level, true));
			}
			if (!doNotGenerateNonGene) {
				tokens.add(new FormulaTerm(nonGene, level, false));
			}
			for (MathmlNode child : node) {
				toTokens(child, level + 1, tokens);
			}
		}
	}

	private void rebuild(MathmlNode node) {
		if (node.isLeaf()) {
			return;
		}
		if (isNaiveNode(node)) {
			return;
		}
		List<MathmlNode> operators = new ArrayList<MathmlNode>();
		List<MathmlNode> operands = new ArrayList<MathmlNode>();
		for (MathmlNode child : node) {
			if (isOperator(node)) {
				while (!operators.isEmpty()) {
					MathmlNode top = operators.get(operators.size() - 1);
					if (isLess(top, child)) {
						operators.add(child);
						break;
					} else {
						if (isBinaryOperator(top)) {
							if (operands.isEmpty()) {
								MathmlTag operator = newOperatorNode(top);
								operands.add(operator);
							} else if (operands.size() == 1) {
								MathmlNode topOperand = operands.get(operands
										.size() - 1);
								MathmlTag operator = newOperatorNode(top);
								operator.addChild(topOperand);
								topOperand.setParent(operator);
								operands.add(operator);
							} else if (operands.size() > 1) {
								MathmlNode topOperand1 = operands.get(operands
										.size() - 1);
								MathmlNode topOperand2 = operands.get(operands
										.size() - 2);
								MathmlTag operator = newOperatorNode(top);
								operator.addChild(topOperand1);
								operator.addChild(topOperand2);
								topOperand1.setParent(operator);
								topOperand2.setParent(operator);
								operands.add(operator);
							}
						} else {
							// unary
							if (operands.isEmpty()) {
								MathmlTag operator = newOperatorNode(top);
								operands.add(operator);
							} else {
								MathmlNode topOperand = operands.get(operands
										.size() - 1);
								MathmlTag operator = newOperatorNode(child);
								operator.addChild(topOperand);
								topOperand.setParent(operator);
								operands.add(operator);
							}
						}
					}
				}
			} else {
				operands.add(child);
			}
		}
	}

	private MathmlTag newOperatorNode(MathmlNode node) {
		return new MathmlTag(node.getParent(), getNaiveNodeLabel(node));
	}

	private boolean isBinaryOperator(MathmlNode node) {
		return true;
	}

	private boolean isLess(MathmlNode node1, MathmlNode node2) {
		return true;
	}

	private int getOperatorPriority(MathmlNode operator) {
		String sign = getNaiveNodeLabel(operator);
		if (sign.equals("-")) {
			// unary - ?? how
			return 2;
		}
		if (sign.equals("^")) {
			return 3;
		}
		if (sign.equals("*") || sign.equals("/") || sign.equals("%")) {
			return 4;
		}
		if (sign.equals("+") || sign.equals("-")) {
			return 5;
		}
		if (sign.equals("<") || sign.equals("<=") || sign.equals(">")
				|| sign.equals(">=")) {
			return 6;
		}
		if (sign.equals("==") || sign.equals("!=")) {
			return 7;
		}
		if (sign.equals("&&")) {
			return 8;
		}
		if (sign.equals("||")) {
			return 9;
		}
		if (sign.equals("=")) {
			return 10;
		}
		return 1;
	}

	private String getNaiveNodeLabel(MathmlNode node) {
		return node.iterator().next().getLabel();
	}

	private boolean isNaiveNode(MathmlNode node) {
		return !node.isLeaf()
				&& (node.getLabel().equals("mi")
						|| node.getLabel().equals("mo") || node.getLabel()
						.equals("mn"));
	}

	private boolean isOperator(MathmlNode node) {
		return !node.isLeaf() && node.getLabel().equals("mo");
	}

	private String getTagString(MathmlNode node) {
		return "<" + node.getLabel() + "></" + node.getLabel() + ">";
	}

	private String getTagString(MathmlNode node, String content) {
		return "<" + node.getLabel() + ">" + content + "</" + node.getLabel()
				+ ">";
	}

	private String getChildrenTagString(MathmlNode node) {
		StringBuilder builder = new StringBuilder();
		for (MathmlNode child : node) {
			if (child.isText()) {
				builder.append(child.getLabel());
			} else if (isOperator(child)) {
				builder.append(child.toString());
			} else {
				builder.append(getTagString(child));
			}
		}
		return builder.toString();
	}

	private String getChildrenString(MathmlNode node) {
		StringBuilder builder = new StringBuilder();
		for (MathmlNode child : node) {
			builder.append(child.toString());
		}
		return builder.toString();
	}
}

File: src/main/scala/com/linxiaoyan/wikimirs/MathmlTokenizer.scala
package com.linxiaoyan.wikimirs

import org.apache.commons.lang3.StringUtils
import scala.collection.mutable.ListBuffer

trait MathmlNode extends Ordered[MathmlNode] {

  def parent: MathmlNode

  def children: List[MathmlNode]
  def compare(that: MathmlNode) = {
    this.toString.compare(that.toString)
  }
}

class MathmlTag(val parent: MathmlTag, val tag: String,
  _children: ListBuffer[MathmlNode] = ListBuffer[MathmlNode]()) extends MathmlNode {

  def children: List[MathmlNode] = _children.toList

  def addChild(child: MathmlNode) {
    _children += child
  }

  val tagString = s"<${tag}></${tag}>"

  def tagString(content: String): String = s"<${tag}>${content}</${tag}>"

  override def toString() =
    s"<${tag}>${children.mkString}</${tag}>";
}

class Text(val parent: MathmlTag, val text: String) extends MathmlNode {
  override val toString = text
  val children = Nil
}

//class MSpace extends MathmlNode {
//
//  override val label: String = "mspace"
//  override val toString = "<mspace/>"
//  override val size = 0
//}

trait TagText extends MathmlNode {
  // Do not use parent
  def parent: MathmlNode = null

  def tag: String
  def text: String
  val tagString = s"<${tag}></${tag}>"
  override val toString = s"<${tag}>${text}</${tag}>"
}

abstract class NaiveNode(val text: String, val tag: String) extends TagText {
  val children = Nil
}

class MText(text: String) extends NaiveNode(text, "mtext") {
}

class MI(text: String) extends NaiveNode(text, "mi") {
}

class MN(text: String) extends NaiveNode(text, "mn") {
}

class MO(val text: String, val children: List[MathmlNode] = List[MathmlNode]()) extends TagText {
  val tag = "mo"
  var operantSize = 0
  override val toString = s"<mo o='${text}'>${children.mkString}</mo>"
  override val tagString = s"<mo o='${text}'></mo>"
  def tagString(content: String): String = s"<mo o='${text}'>${content}</mo>"
}

class MS(text: String) extends NaiveNode(text, "ms") {
}

class MathmlTokenizer(val enableGen: Boolean = true) {

  def toTokens(node: MathmlNode, tokens: ListBuffer[FormulaTerm]) {
    toTokens(node, 1, tokens);
  }

  def toTokens(node: MathmlNode, level: Int, tokens: ListBuffer[FormulaTerm]): Unit = node match {
    case node: NaiveNode => {
      val text = node.text
      if (text.length > 1) {
        tokens += new FormulaTerm(node.toString, level, false)
      }
    }
    case node: Text => {
      if (node.text.length <= 1) {
        return
      }
      tokens += new FormulaTerm(node.toString, level, false)
    }
    case node: MO => {
      val generalized = node.tagString(getChildrenTagString(node))
      val nonGeneralized = node.tagString(getChildrenString(node))
      if (!generalized.equals(nonGeneralized)) {
        if (enableGen) {
          tokens += new FormulaTerm(generalized, level, true)
        }
      }
      tokens += new FormulaTerm(nonGeneralized, level, false)
      node.children.foreach {
        toTokens(_, level + 1, tokens);
      }
    }
    case node: MathmlTag => {
      var generalized = getChildrenTagString(node);
      var nonGeneralized = getChildrenString(node);
      var skipThisNode = false;
      if (node.tag.equals("math")
        || node.tag.equals("mrow")) {
        if (node.children.size == 1) {
          // <math><mi>a</mi></math> 
          // nonGeneralized => <mi>a</mi>
          // already generated
          skipThisNode = true
        }
      } else {
        generalized = node.tagString(generalized)
        nonGeneralized = node.tagString(nonGeneralized)
      }
      if (!skipThisNode) {
        if (!generalized.equals(nonGeneralized)) {
          if (enableGen) {
            tokens += new FormulaTerm(generalized, level, true)
          }
        }
        tokens += new FormulaTerm(nonGeneralized, level, false);
      }
      node.children.foreach {
        toTokens(_, level + 1, tokens);
      }
    }
  }

  private def getChildrenTagString(node: MathmlNode): String = {
    node.children.map {
      case o: Text => o.text
      case o: TagText => o.tagString
      case o: MathmlTag => o.tagString
    }.mkString
  }

  private def getChildrenString(node: MathmlNode) = {
    node.children.map(_.toString).mkString
  }
}
